design_notes/v2.md

```diff
@@ -0,0 +1,176 @@
+
+```python
+class FileReader(Protocol):
+    """
+    Protocol for file operations for downstream libraries
+    """
+
+    def read(self, size: int, /) -> bytes:
+        ...
+
+    def seek(self, offset: int, whence: int = 0, /):
+        ...
+
+    def tell(self) -> int:
+        ...
+
+    async get_range_async(
+        self,
+        path: str,
+        *,
+        start: int,
+        end: int | None = None,
+        length: int | None = None
+    ) -> Bytes :
+        ...
+
+    async get_async(
+        self,
+        path: str,
+        *,
+        options: int | None = None,
+        length: int | None = None
+    ) -> GetResult :
+        ...
+
+```
+
+
+```python
+class VirtualZarrBackend(protocol):
+    """
+    Protocol for creating a virtual zarr dataset from a file
+    """
+
+    @staticmethod
+    def open_virtual_store(
+        filepath: str,
+        *,
+        virtual_backend_kwargs: dict | None = None,
+        reader_options: ObspecReaderOptions | dict | None = None,
+    ) → VirtualStore:
+        ...
+```
+
+> **Question:** How do we define this protocol to allow for strict typing for some readers (e.g., obstore-backed) and not others (e.g., fsspec-backed)?
+
+> **Note:** Could ObspectReaderOptions be typed as follows?
+
+```python
+ObspecReaderOptions(TypedDict):
+    s3_config: S3Config | None = None
+    azure_config: AzureConfig | None = None
+    gcs_config: GCSConfig | None = None
+    client_options: NotRequired[ClientConfig]
+    retry_config: NotRequired[RetryConfig]
+    credential_provider: NotRequired[Callable]
+```
+## Top-level API functions
+
+- `open_virtual_dataset` - construct a Xarray Dataset with NumPy and/or ManifestArray-backed variables from a file
+- `open_virtual_mfdataset` - construct a Xarray Dataset with NumPy and/or ManifestArray-backed variables from many files
+
+> **NOTE**: Here the construction of a virtual dataset from a manifest store would happen in the top-level function rather than the backends, which would need to implement `open_virtual_store` rather than `open_virtual_dataset`
+
+```python
+ def open_virtual_dataset(
+    filepath: str,
+    *,
+    cache: bool = False,
+    drop_variables: Iterable[str] | None = None,
+    loadable_variables: Iterable[str] | None = None,
+    decode_times: bool | None = None,
+    cftime_variables: Iterable[str] | None = None,
+    indexes: Mapping[str, Index] | None = None,
+    virtual_backend_kwargs: dict | None = None,
+    reader_options: ObspecReaderOptions | dict | None = None,
+    backend: type[VirtualBackend] | None = None
+) → Dataset
+    """
+    Method for creating virtual Xarray dataset, only implemented in top level API
+    """
+    # Guess backend if not provided based on filepath or magic bytes
+    if cache:
+        orig_fp = filepath
+        filepath = cache_file(orig_fp, reader_options)
+    if not backend:
+        backend = guess_backend(filepath, reader_options)
+    backend.open_virtual_store(
```

Comment:
`backend.open_virtual_store` should just be a function IMO (see other comments above). There's no state to keep track of on the `backend` object, and it only needs one method, so it doesn't need to be a class.

Reply:
how do we standardize the function signatures on backends if it's just a function? I only know how to do this using ABCs or protocols TBH, which is why I left it as a protocol.

Reply:
Define a [callback protocol](https://typing.python.org/en/latest/spec/callables.html#callback-protocols).

Reply:
Thanks so much, Chuck!